### PR TITLE
19016 double extension point v22

### PR DESCRIPTION
**Pull request by chrom**
Fix double extension point of product conversion service when generating simple products through the wizard in product page.
https://github.com/magento/magento2/issues/19016

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=19042) <br/>All committers have signed the CLA.

**Comment by magento-engcom-team** [comment]
Hi @chrom. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me $VERSION instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

File: app/code/Magento/ConfigurableProduct/Model/Product/VariationHandler.php [hunk 1]
```diff
@@ -5,12 +5,19 @@
...
+use Magento\CatalogInventory\Api\StockConfigurationInterface;
+use Magento\Eav\Model\Entity\Attribute\SetFactory;
+use Magento\Eav\Model\EntityFactory;
 use Magento\Framework\Exception\LocalizedException;
+use Magento\ConfigurableProduct\Model\Product\ChildrenWeightResolver;
 
 /**
  * Variation Handler
  * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
+ *
```
Comment (orlangur) [inline app/code/Magento/ConfigurableProduct/Model/Product/VariationHandler.php:20]: Such changes must be reverted.

File: app/code/Magento/ConfigurableProduct/Model/Product/VariationHandler.php [hunk 2]
```diff
@@ -47,49 +54,59 @@ class VariationHandler
...
      * @return array
      * @throws \Magento\Framework\Exception\LocalizedException
      */
     public function generateSimpleProducts($parentProduct, $productsData)
     {
         $generatedProductIds = [];
-        $this->attributes = null;
-        $productsData = $this->duplicateImagesForVariations($productsData);
+        $this->attributes    = null;
+        $productsData        = $this->duplicateImagesForVariations($productsData);
```
Comment (orlangur) [inline app/code/Magento/ConfigurableProduct/Model/Product/VariationHandler.php:109]: Such changes must be reverted.

File: app/code/Magento/ConfigurableProduct/Model/Product/VariationHandler.php [hunk 3]
```diff
@@ -47,49 +54,59 @@ class VariationHandler
...
     public function __construct(
         Type\Configurable $configurableProduct,
-        \Magento\Eav\Model\Entity\Attribute\SetFactory $attributeSetFactory,
-        \Magento\Eav\Model\EntityFactory $entityFactory,
-        \Magento\Catalog\Model\ProductFactory $productFactory,
-        \Magento\CatalogInventory\Api\StockConfigurationInterface $stockConfiguration,
-        \Magento\Catalog\Model\Product\Gallery\Processor $mediaGalleryProcessor
+        SetFactory $attributeSetFactory,
+        EntityFactory $entityFactory,
+        ProductFactory $productFactory,
```
Comment (orlangur) [inline app/code/Magento/ConfigurableProduct/Model/Product/VariationHandler.php:81]: No need to revert but generally please avoid such changes mixed with change of logic. We will get rid of FQCN at some point but it should be a separate effort.

**Review by orlangur — CHANGES_REQUESTED** [review]
I see three lines of code encapsulated in a new class and a lot of unrelated changes. Where are the changes related to `TypeTransition`? I don't see extension points decreased.

Currently all changes needs to be applied to `2.3-develop` first.

**Comment by chrom** [comment]
@orlangur This makes it possible to disable the autoconversion of products using the plugin. We discussed this with @maghamed  and @naydav.

**Comment by orlangur** [comment]
@chrom,

> makes it possible to disable the autoconversion of products using the plugin

When this may be needed? Please provide a complete use case. What's the problem with specifying weight, for instance?

**Comment by chrom** [comment]
@orlangur When the customer needs to disable the autoconversion of the product. As an example, we have such a client who asked to turn off autoconversion as it does not have downloaded or virtual products and due to the human factor - when managers forgot to put weight, autoconversion occurred.

**Comment by orlangur** [comment]
@chrom just make weight a mandatory field then.

> when managers forgot to put weight, autoconversion occurred

Proper weight may be added afterwards.

**Comment by chrom** [comment]
when generating simple products through the wizard, default attribute
values are not set

чт, 8 нояб. 2018 г. в 16:18, Vlad Veselov <notifications@github.com>:

**Comment by orlangur** [comment]
@chrom,

> when generating simple products through the wizard, default attribute
values are not set

I would rather treat it as a bug and fix.

**Comment by chrom** [comment]
Yes, but it still does not allow you to disable the conversion - say, based
on dynamic rules

вт, 13 нояб. 2018 г. в 22:54, Vlad Veselov <notifications@github.com>:

**Comment by orlangur** [comment]
> based on dynamic rules

@chrom the whole idea was that merchant should not care about product type at all, conversion occurs transparently and automatically. Please provide scenario where it is really essential whether configurable product child products are virtual or simple.

**Comment by orlangur** [comment]
Closing as this PR does not seem to achieve goal described in issue: having a single product type conversion extension point.

Feel free to reach me out anytime later if you wish to continue work on this pull request and it will be reopened. Thanks for collaboration!

**Pull request closed without merging**